File: products/product_repository.py
from sqlalchemy import select

from products.product_db import db, Product


class ProductRepository:

    @staticmethod
    def get_by_name(name: str) -> Product:
        return db.session.query(Product).filter(Product.product_name == name).one_or_none()

    @staticmethod
    def save(product: Product):
        """
        Insert a new product into the product table with provided information as columns.
        """

        # If product already exists then update product, else insert new product
        # a product is a new product if it does not have a product_id/product_id = None
        if product.product_id is None:
            # add is called in order to have access to product.product_id
            db.session.add(product)
        else:
            db.session.merge(product)
        db.session.commit()

    """def delete():"""

File: Spendy/Spendy.Shared/Repositories/ProductRepository.cs
using Spendy.Shared.Data;
using Spendy.Shared.Models;

namespace Spendy.Shared.Repositories;

public interface IProductRepository
{
    public void Save(Product product);
    public void Update(Product product);
    public Product? GetByName(string productName);
    public List<Product> GetAllProducts();
}

public class ProductRepository: IProductRepository
{
    private readonly SpendyContext _context;

    public ProductRepository(SpendyContext context)
    {
        _context = context;
    }

    public List<Product> GetAllProducts()
    {
        var products = _context.Products
            .ToList();
        return products;
    }

    // Return product with given name or return null is product does not exist
    public Product? GetByName(string productName)
    {
        var product = _context.Products
            .FirstOrDefault(p => p.ProductName == productName);
        return product;
    }
    // save a new product
    public void Save(Product product)
    {
        Console.WriteLine("Save Product");
        _context.Add(product);
    }
    // if product exists already, update it
    public void Update(Product product)
    {
        Console.WriteLine();
        _context.Update(product);
    }
    // delete a product
    public void Delete()
    {
        Console.WriteLine("Delete Product");
    }
}